### ISSUE TITLE
When I am not logged in as an admin, I can still navigate to some admin routes instead of getting redirect to the login patge

### ISSUE THREAD

**Issue by MAF27** [freeCodeCamp/food-bank-app#34]
/#!/admin/volunteers
/#!/admin/donors
/#!/admin/questionnaires

**Comment by kenjiO** [comment]
OK if I try to take this one on?

**Comment by DealPete** [comment]
by all means

**Comment by kenjiO** [comment]
I found a way to fix it using code from the admin-customer controller.
https://github.com/kenjiO/food-bank-app/commit/6bf8176530d4e6c16af184a7beb0ba0b394ab540
I changed the logic so it still redirects to root if user or user.roles is not set which is the case when no one is logged on.

I thought another way to do it is use the ui-router resolve property similar to how it is used on [this page](https://github.com/aaronksaunders/hu1/wiki/4.3-An-Approach-to-Authentication-Flow-with-UI-Router).  This way it can check that an admin user is logged on before loading the controller and view.  But this would mean a lot more changes to the code.

**Comment by DealPete** [comment]
Hi Kenji,

I really like the ui-router based authentication. It would be better than the hodge-podge we have currently. Like you say, though, it would be an effort to implement, and the first method would adequately address this issue. We'll leave it up to you which to go with.

**Comment by kenjiO** [comment]
The above commit is an idea of how I got the ui-router to authenticate an admin route so you can keep it out of the controllers and make the code more DRY.

I thought the best place to put the method that checks for an admin is in the `users` module `Authentication` factory.  But the ui-router code is in app.config sections, which don't have access to factories.  So I changed the `Authentication` factory to be a Provider which can also be used like a factory.  Now in any of the ui-router code you can just include the `AuthenticationProvider` dependency. Then for a route that should be restricted to admins add
```javascript
resolve: {
    CurrentUser: AuthenticationProvider.requireAdminUser
}
```

The `CurrentUser` name is arbitrary and is not used for anything yet, but it does get injected into the controller and gives access to the user object returned when the `requireAdminUser()` promise is resolved.

So far I only added this functionality to the 'admin/customers' route.  If you want to go ahead with it I can add it to all the other admin routes and do a PR.
Here is an idea of how I got the ui-router to authenticate an admin route so you can keep it out of the controllers and make the code more DRY.

I thought the best place to put the method that checks for an admin is in the `users` module `Authentication` factory.  But the ui-router code is in app.config sections, which don't have access to factories.  So I changed the `Authentication` factory to be a Provider which can also be used like a factory.  Now in any of the ui-router code you can just include the `AuthenticationProvider` dependency. Then for a route that should be restricted to admins add
```javascript
resolve: {
    CurrentUser: AuthenticationProvider.requireAdminUser
}
```

The `CurrentUser` name is arbitrary and is not used for anything yet, but it does get injected into the controller and gives access to the user object returned when the `requireAdminUser()` promise is resolved.

So far I only added this functionality to the 'admin/customers' route.  If you want to go ahead with it I can add it to all the other admin routes and do a PR.

**Comment by kenjiO** [comment]
Sorry I was trying to fix some extra whitespace in commit b5dd628 and inadvertently added fa41ba7 to this discussion. The only difference between the two is some whitespace.

**Comment by DealPete** [comment]
This looks like a great improvement, thank you.
